// @myfe/cat-datepicker/dist/pickerfooter.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var PickerFooter = (function (_Component) {
    _inherits(PickerFooter, _Component);

    function PickerFooter(props) {
        var _this = this;

        _classCallCheck(this, PickerFooter);

        _get(Object.getPrototypeOf(PickerFooter.prototype), 'constructor', this).call(this, props);

        this.handleChooseNow = function () {
            var date = new Date();
            _this.props.dateSwitch(date, 'date');
        };

        this.handleClick = function () {
            _this.props.dateSwitch(_this.props.selectedDate, 'date');
        };
    }

    _createClass(PickerFooter, [{
        key: 'renderNow',
        value: function renderNow() {
            if (!this.props.showNow) {
                return null;
            } else {
                var word = this.props.timeOnly === true ? '现在' : '今天';
                return _react2['default'].createElement(
                    'button',
                    {
                        className: 'btn btn-link btn-block',
                        onClick: this.handleChooseNow,
                        type: 'button'
                    },
                    word
                );
            }
        }
    }, {
        key: 'renderConfirm',
        value: function renderConfirm() {
            if (this.props.dateOnly === true && this.props.timeOnly !== true) {
                return null;
            } else {
                return _react2['default'].createElement(
                    'button',
                    {
                        className: 'btn btn-default btn-block',
                        onClick: this.handleClick,
                        type: 'button'
                    },
                    '确定'
                );
            }
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2['default'].createElement(
                'div',
                { className: 'cat-calendar-cell-footer clearfix' },
                this.renderNow(),
                this.renderConfirm()
            );
        }
    }]);

    return PickerFooter;
})(_react.Component);

exports['default'] = PickerFooter;
module.exports = exports['default'];
