// @myfe/cat-sweetalert/dist/sweet.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _myfeCatDialog = require('@myfe/cat-dialog');

var _myfeCatDialog2 = _interopRequireDefault(_myfeCatDialog);

var _alert = require('./alert');

var _alert2 = _interopRequireDefault(_alert);

var _confirm = require('./confirm');

var _confirm2 = _interopRequireDefault(_confirm);

var SweetAlert = (function (_Component) {
    _inherits(SweetAlert, _Component);

    function SweetAlert() {
        _classCallCheck(this, SweetAlert);

        _get(Object.getPrototypeOf(SweetAlert.prototype), 'constructor', this).apply(this, arguments);

        this.state = {
            show: false
        };
    }

    _createClass(SweetAlert, [{
        key: 'show',
        value: function show(type, config, callback) {
            this.setState({
                show: true,
                type: type,
                config: config,
                callback: callback
            });
        }
    }, {
        key: '_close',
        value: function _close() {
            this.setState({
                show: false
            });
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2['default'].createElement(
                _myfeCatDialog2['default'],
                {
                    show: this.state.show,
                    toBody: false,
                    onClose: this._close.bind(this),
                    noButtons: true,
                    title: '',
                    autoClose: false
                },
                this._renderContent()
            );
        }
    }, {
        key: '_renderContent',
        value: function _renderContent() {
            switch (this.state.type) {
                case 'alert':
                    return _react2['default'].createElement(_alert2['default'], {
                        config: this.state.config,
                        close: this._close.bind(this)
                    });
                    break;
                case 'confirm':
                    return _react2['default'].createElement(_confirm2['default'], {
                        config: this.state.config,
                        callback: this.state.callback,
                        close: this._close.bind(this)
                    });
                    break;
                case 'prompt':
                    return _react2['default'].createElement(_confirm2['default'], {
                        config: this.state.config,
                        callback: this.state.callback,
                        close: this._close.bind(this),
                        isPrompt: true
                    });
                    break;
            }
            return null;
        }
    }]);

    return SweetAlert;
})(_react.Component);

SweetAlert.newInstance = function (type) {
    var div = document.createElement('div');
    div.className = 'my-sweetalert-container';
    document.body.appendChild(div);
    var sweet = _reactDom2['default'].render(_react2['default'].createElement(SweetAlert, null), div);
    return {
        show: function show(type, config, callback) {
            sweet.show(type, config, callback);
        },
        component: sweet
    };
};

exports['default'] = SweetAlert;
module.exports = exports['default'];
