// @myfe/cat-grid/dist/grid/gridth.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var GridTh = (function (_Component) {
    _inherits(GridTh, _Component);

    _createClass(GridTh, null, [{
        key: 'propTypes',
        value: {
            column: _react2['default'].PropTypes.object,
            orderFunc: _react2['default'].PropTypes.object
        },
        enumerable: true
    }, {
        key: 'defaultProps',
        value: {
            column: {},
            orderFunc: {
                key: '',
                func: function func() {},
                forward: true
            }
        },
        enumerable: true
    }]);

    function GridTh(props) {
        var _this = this;

        _classCallCheck(this, GridTh);

        _get(Object.getPrototypeOf(GridTh.prototype), 'constructor', this).call(this, props);

        this._handleClick = function (event) {
            var column = _this.props.column;

            _this.props.handleOrder(column.name, column.orderBy);
        };
    }

    _createClass(GridTh, [{
        key: 'render',
        value: function render() {
            var _props = this.props;
            var column = _props.column;
            var orderFunc = _props.orderFunc;
            var self = this;

            return _react2['default'].createElement(
                'th',
                null,
                column.label,
                (function () {
                    if (column.orderBy) {
                        var classes = orderFunc.key !== column.name ? 'sort' : orderFunc.forward ? 'sort-by-attributes' : 'sort-by-attributes-alt';
                        return _react2['default'].createElement('i', {
                            className: 'glyphicon glyphicon-' + classes,
                            onClick: self._handleClick
                        });
                    }
                })()
            );
        }
    }]);

    return GridTh;
})(_react.Component);

exports['default'] = GridTh;
module.exports = exports['default'];
