// @myfe/cat-grid/dist/grid/gridhead.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _gridthJs = require('./gridth.js');

var _gridthJs2 = _interopRequireDefault(_gridthJs);

var GridHead = (function (_Component) {
    _inherits(GridHead, _Component);

    _createClass(GridHead, null, [{
        key: 'propTypes',
        value: {
            columns: _react2['default'].PropTypes.array,
            enableSelection: _react2['default'].PropTypes.bool,
            selectAll: _react2['default'].PropTypes.bool,
            myStyle: _react2['default'].PropTypes.string,
            order: _react2['default'].PropTypes.bool,
            orderFunc: _react2['default'].PropTypes.object
        },
        enumerable: true
    }, {
        key: 'defaultProps',
        value: {
            columns: [],
            enableSelection: false,
            selectAll: false,
            myStyle: 'active',
            order: false,
            orderFunc: {
                key: '',
                func: function func() {},
                forward: true
            }
        },
        enumerable: true
    }]);

    function GridHead(props) {
        var _this = this;

        _classCallCheck(this, GridHead);

        _get(Object.getPrototypeOf(GridHead.prototype), 'constructor', this).call(this, props);

        this._handleClick = function (event) {
            _this.props.onSelect('-1');
        };

        this._handleOrder = function (newKey, func) {
            var orderFunc = _this.props.orderFunc;

            if (orderFunc.key !== newKey) {
                orderFunc = {
                    key: newKey,
                    func: func,
                    forward: true
                };
            } else if (orderFunc.forward) {
                orderFunc.forward = false;
            } else {
                orderFunc = {
                    key: '',
                    func: function func() {},
                    forward: true
                };
            }

            _this.props.updateOrder(orderFunc);
        };
    }

    _createClass(GridHead, [{
        key: 'renderSelection',
        value: function renderSelection() {
            var _props = this.props;
            var enableSelection = _props.enableSelection;
            var selectAll = _props.selectAll;
            var iconClass = selectAll ? 'glyphicon glyphicon-check' : 'glyphicon glyphicon-unchecked';
            if (!enableSelection) {
                return null;
            } else {
                return _react2['default'].createElement(
                    'th',
                    { onClick: this._handleClick, className: 'table-th-hover' },
                    _react2['default'].createElement('i', { className: iconClass })
                );
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _props2 = this.props;
            var columns = _props2.columns;
            var orderFunc = _props2.orderFunc;
            var myStyle = _props2.myStyle;
            var self = this;

            return _react2['default'].createElement(
                'thead',
                null,
                _react2['default'].createElement(
                    'tr',
                    { className: myStyle },
                    this.renderSelection(),
                    _lodash2['default'].map(columns, function (column) {
                        return _react2['default'].createElement(_gridthJs2['default'], {
                            key: column.label,
                            column: column,
                            orderFunc: orderFunc,
                            handleOrder: self._handleOrder
                        });
                    })
                )
            );
        }
    }]);

    return GridHead;
})(_react.Component);

;

exports['default'] = GridHead;
module.exports = exports['default'];
