// @myfe/cat-grid/dist/nav/navfooter.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _myfeCatPagination = require('@myfe/cat-pagination');

var _myfeCatPagination2 = _interopRequireDefault(_myfeCatPagination);

var NavFooter = (function (_Component) {
    _inherits(NavFooter, _Component);

    function NavFooter() {
        var _this = this;

        _classCallCheck(this, NavFooter);

        _get(Object.getPrototypeOf(NavFooter.prototype), 'constructor', this).apply(this, arguments);

        this._updatePage = function (newPage) {
            _this.props.update(newPage);
        };
    }

    _createClass(NavFooter, [{
        key: 'render',
        value: function render() {
            var _props = this.props;
            var prefixName = _props.prefixName;
            var offset = _props.offset;
            var limit = _props.limit;
            var number = _props.number;
            var total = _props.total;
            var paginationStyle = _props.paginationStyle;
            var totalPage = total && limit ? Math.ceil(total / limit) : -1;

            return _react2['default'].createElement(
                'nav',
                { className: prefixName + '-table-footer clearfix' },
                _react2['default'].createElement(
                    'div',
                    { className: prefixName + '-table-footnote pull-left' },
                    (total !== 0 || number !== 0) && _react2['default'].createElement(
                        'span',
                        null,
                        ' 共',
                        total !== 0 ? total : number,
                        '条数据  '
                    ),
                    number !== 0 && offset !== -1 && _react2['default'].createElement(
                        'span',
                        null,
                        '当前显示',
                        offset + 1,
                        '-',
                        offset + number,
                        '条'
                    )
                ),
                offset !== -1 && _react2['default'].createElement(_myfeCatPagination2['default'], {
                    offset: offset / limit,
                    totalPage: totalPage,
                    myStyle: paginationStyle,
                    update: this._updatePage,
                    className: 'pull-right',
                    mySize: 'small'
                })
            );
        }
    }], [{
        key: 'propTypes',
        value: {
            prefixName: _react2['default'].PropTypes.string,
            offset: _react2['default'].PropTypes.number,
            limit: _react2['default'].PropTypes.number,
            number: _react2['default'].PropTypes.number,
            total: _react2['default'].PropTypes.number,
            paginationStyle: _react2['default'].PropTypes.string
        },
        enumerable: true
    }, {
        key: 'defaultProps',
        value: {
            prefixName: 'cat',
            offset: -1,
            limit: 1,
            number: 0,
            total: 0,
            paginationStyle: 'omitted'
        },
        enumerable: true
    }]);

    return NavFooter;
})(_react.Component);

exports['default'] = NavFooter;
module.exports = exports['default'];
