// @myfe/cat-message/dist/message.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _notification = require('./notification');

var _notification2 = _interopRequireDefault(_notification);

function addNotice(content, settings, type) {
    var container = undefined,
        notice = {
        content: content,
        type: type
    };

    // TODO: 我们需要规范报错机制，如果用户使用api错误，利用catch-throw给warning
    if (typeof settings === 'number') {
        notice.duration = settings;
    } else if (typeof settings === 'object') {
        if (!settings.nodeType) {
            notice.duration = settings.duration;
            container = settings.container;
        } else {
            container = settings;
        }
    }

    _notification2['default'].instance(container).add(notice);
}

exports['default'] = {
    info: function info(content, settings) {
        addNotice(content, settings, 'info');
    },
    success: function success(content, settings) {
        addNotice(content, settings, 'success');
    },
    warning: function warning(content, settings) {
        addNotice(content, settings, 'warning');
    },
    danger: function danger(content, settings) {
        addNotice(content, settings, 'danger');
    }
};
module.exports = exports['default'];
