// @myfe/cat/src/components/countdown.js
import React, { Component } from 'react';
import Button from '@myfe/cat-button';

export default class CountDown extends Component {

    static propTypes = {
        count: React.PropTypes.number
    }

    static defaultProps = {
        count: 5
    }

    state = {
        disabled: false
    }

    shouldComponentUpdate(nextProps, nextState) {
        return !nextProps.disabled;
    }

    handleClick() {
        let { count } = this.props;

        if (!this.state.disabled) {
            this.props.onCount();
            this.setState({
                disabled: true
            });
            setTimeout(() => {
                this.setState({
                    disabled: false
                });
            }, count * 1000);
        }
    }

    render() {
        let { count } = this.props;
        let { disabled, content } = this.state;
        content = disabled ? `${count}s后再尝试` : this.props.children;

        return <Button {...this.props} onClick={this::this.handleClick} disabled={disabled}>{content}</Button>;
    }

}
