// @myfe/cat-pagination/dist/pager.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var Pager = (function (_Component) {
    _inherits(Pager, _Component);

    _createClass(Pager, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            this.setState({
                disabled: nextProps.target < 0 || nextProps.totalPage !== -1 && nextProps.target >= nextProps.totalPage
            });
        }
    }], [{
        key: 'propTypes',
        value: {
            page: _react2['default'].PropTypes.string,
            target: _react2['default'].PropTypes.number,
            totalPage: _react2['default'].PropTypes.number
        },
        enumerable: true
    }, {
        key: 'defaultProps',
        value: {
            page: '',
            target: -1,
            totalPage: 0
        },
        enumerable: true
    }]);

    function Pager(props) {
        _classCallCheck(this, Pager);

        _get(Object.getPrototypeOf(Pager.prototype), 'constructor', this).call(this, props);
        this.state = {
            disabled: this.props.target < 0 || this.props.totalPage !== -1 && this.props.target >= this.props.totalPage
        };
    }

    _createClass(Pager, [{
        key: '_handleClick',
        value: function _handleClick(event) {
            event.preventDefault();
            if (!this.state.disabled) {
                this.props.handleClick(this.props.target);
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var disabled = this.state.disabled ? 'disabled' : '',
                active = this.props.target === this.props.offset ? ' active' : '',
                classes = disabled + ' ' + active;

            return _react2['default'].createElement(
                'li',
                { className: classes },
                _react2['default'].createElement(
                    'a',
                    { href: '#', onClick: this._handleClick.bind(this) },
                    this.props.page
                )
            );
        }
    }]);

    return Pager;
})(_react.Component);

exports['default'] = Pager;
module.exports = exports['default'];
