// @myfe/cat-message/src/notice.js
import React, { Component } from 'react';

export default class Notice extends Component {
    static propTypes = {
        duration: React.PropTypes.number,
        content: React.PropTypes.string,
        preFixName: React.PropTypes.string
    }
    static defaultProps = {
        duration: 5,
        content: 'This is a notice',
        preFixName: 'cat'
    }

    constructor(props) {
        super(props);
    }

	close() {
        this.props.onClose();
	}

	clearTimer() {
		if (this.timer) {
			clearTimeout(this.timer);
			this.timer = null;
		}
	}

	componentDidMount() {
        let { duration } = this.props;
		if (duration !== 0) {
			this.timer = setTimeout(() => {
				this.close();
			}, duration * 1000);
		}
	}

	shouldComponentUpdate(nextProps) {
		return false;
	}

	componentWillUnmount() {
		this.clearTimer();
	}

    render() {
		let { type, preFixName } = this.props,
            // TODO: 更新相关的less文件，使用有前缀的类名, 去除没有逻辑的style
            cls = `${preFixName}-alert ${preFixName}-alert-${type}`;

        return (
            <div className={cls}>
				{ this.props.children }
				<i className="close" onClick={this.close.bind(this)}>&times;</i>
            </div>
        );
    }
}
