// @myfe/cat/examples/get-side-info.js
import MENU_DATA from './menu-data';
import _ from 'lodash';

function findCurrent(current, data) {
    let self = this,
        list = [];
    _.find(data, function(chr) {
        if (chr.href === current) {
            list.push({name: chr.name, href: chr.href});
            return true;
        } else if (chr.children) {
            let subList = findCurrent(current, chr.children);
            if (subList.length > 0) {
                list = [{name: chr.name, href: chr.href}].concat(subList);
                return true;
            } else {
                return false;
            }
        }
    });

    return list;
}

function getSideInfo(current) {
    let breads = findCurrent(current, MENU_DATA);
    breads.unshift({name: '首页', href: '/'});
    
    return {
        breads: breads,
        current: current,
        currentMenuData: MENU_DATA,
        channel: {
            name: '营销推广',
            href: '/yinxiao',
            channelList: MENU_DATA
        }
    };
}

export default getSideInfo;
